src/AspNetCore/Src/applicationmanager.cxx

```diff
@@ -124,10 +124,12 @@ APPLICATION_MANAGER::GetApplicationInfo(
         }
 
         *ppApplicationInfo = pApplicationInfo;
-        ReleaseSRWLockExclusive(&m_srwLock);
-        fExclusiveLock = FALSE;
 
+        // Listen to app offline before releasing the lock, which will avoid races here.
         pApplicationInfo->StartMonitoringAppOffline();
+        ReleaseSRWLockExclusive(&m_srwLock);
```

Comment:
how can we make sure after releasing the lock, an appoffline happens immediately and the applicationinfo is not release prematurely

Reply:
Shutdown + dereferencing should avoid this issue. Again I can talk to you tomorrow about this.